src/quest/persistence.py

```diff
@@ -19,45 +19,95 @@ def has_blob(self, key: str) -> bool: ...
 
     def delete_blob(self, key: str): ...
 
+class HistoryNode():
+    def __init__(self):
+        self.prev: HistoryNode = None
+        self.next: HistoryNode = None
+        self.item = None
 
 class PersistentHistory(History):
     def __init__(self, namespace: str, storage: BlobStorage):
         self._namespace = namespace
         self._storage = storage
-        # TODO - use linked list instead of array list
-        # master stores head/tail keys
-        # each blob is item, prev, next
-        # only need to modify blobs for altered nodes
-        # on add/delete, and rarely change master head/tail blob
-        self._items = []
-        self._keys: list[str] = []
+        self._nodes: dict[str, HistoryNode] = {}
+
+        self._head: HistoryNode = None
+        self._tail: HistoryNode = None
 
         if storage.has_blob(namespace):
-            self._keys = storage.read_blob(namespace)
-            for key in self._keys:
-                self._items.append(storage.read_blob(key))
+            keys = storage.read_blob(namespace)
+            for key in keys:
+                self.append(storage.read_blob(key))
 
     def _get_key(self, item: EventRecord) -> str:
         return self._namespace + '.' + md5((item['timestamp'] + item['step_id'] + item['type']).encode()).hexdigest()
 
+    def _insert_node(self, nodeKey: str, item: Blob):
+        """Inserts blob into the linked list of HistoryNodes"""
+        newNode: HistoryNode = HistoryNode()
+        newNode.item = item
+        newNode.prev = self._tail
+
+        if(self._tail != None):
+            self._tail.next = newNode
+        else:
+            self._head = newNode
+
+        self._tail = newNode
+
+        self._nodes[nodeKey] = newNode
+
+    def _remove_node(self, nodeKey: str):
+        # TODO: should I be key checking here first? Probably
+        toDelete: HistoryNode = self._nodes[nodeKey]
+
+        if toDelete.prev is not None:
+            toDelete.prev.next = toDelete.next
+
+        if toDelete.next is not None:
+            toDelete.next.prev = toDelete.prev
+
+        if toDelete == self._head:
```

Comment:
If you are doing object instance comparison, instead of value comparison, use `is`:
`if toDelete.next is self._head`
`is` compares memory addresses (i.e. are these the *same* object as opposed to are these *equivalent* objects).